### DISCUSSION TITLE
Measure a channel

### DISCUSSION THREAD

**Discussion by pfeatherstone** [jgaeddert/liquid-dsp#300] [How do I....?]
How can I measure channel impairments given some hardware setup?
So I have a transmitter, receiver, antennae and I'm transmitting a simple BPSK over the air.
What types of signals and what algorithms can I run to measure the channel. For example:
* AWGN
* Phase jitter
* Time jitter
* Fading
* Doppler
* Multipath
* etc
Are there DSP methods for doing this?
Or do I have to make an assumption on the channel impairment algorithms, collect all their parameters (un-initialized), then run some global optimizer using some RF captures as ground truth data?
Any ideas?

**Comment by jgaeddert** [comment]
Great questions. Currently the `symtrack` family of objects perform correction for many of these channel impairments, but don't provide measurements for these directly. I'd be interested in re-vectoring that object with new APIs that provide access to those measurements.

A few questions:

  * is your BPSK signal continuous or bursty?
  * does your BPSK signal have pilots or a synchronization word? or just random bits?

**Reply by pfeatherstone** [reply]
At the moment the BPSK is just continuous and i'm transmitting random bits. But i don't mind transmitting anything. If you tell me that actually in order to do this well you need to transmit a CW signal, then i'll do that. I just need a set of tests that allow me to measure all these different channel properties.

**Reply by jgaeddert** [reply]
Modulated data work better for channel measurements than CW. Right now the [symtrack api](https://liquidsdr.org/api/symtrack_cccf/) doesn't have methods for getting offset estimates. I imagine we would start with things like:

```c
// get the frequency offset in radians per sample
float symtrack_cccf_get_nco_frequency(symtrack_cccf _q);
```

etc.

**Reply by pfeatherstone** [reply]
I'm also interested in measuring the phase noise spectrum of an SDR and estimate flicker frequency and corner frequency.

**Reply by jgaeddert** [reply]
Do you have a reference for those? The phase noise would actually be better estimated with a tone, despite my comment earlier.

**Reply by pfeatherstone** [reply]
Those things probably have different names in different books. I'll dig some references out. But you see my point. Depending on the system impairment or hardware impairment you're trying to measure, different signals, tones or whatever are needed. I'm keen to know what signals to use and what are the tests. Cheers

**Reply by jgaeddert** [reply]
@pfeatherstone did you ever find those references?

**Reply by pfeatherstone** [reply]
Oh I completely forgot. I'll try remember.

**Reply by pfeatherstone** [reply]
https://en.m.wikipedia.org/wiki/Leeson%27s_equation talks of corner frequency. I'll try find a source that mentions flicker frequency. I'm pretty sure I'm not going mad

**Reply by pfeatherstone** [reply]
This also talks about this https://www.ieee.li/pdf/essay/phase_noise_basics.pdf
